### ISSUE TITLE
Changing duration of clips with keyframe filters does not adjust filters

### ISSUE THREAD

**Issue by alexrudd2** [mltframework/shotcut#750]
Linux kol 4.15.0-52-generic #56-Ubuntu SMP Tue Jun 4 22:49:08 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux

Shotcut 19.06.15 / Qt 5.2.4.

Observed behaviour:
Add a filter to a video clip (or picture)
Click keyframes next to a track parameter.  Observe dot at red keyframe in keyframe view.
Change the clip duration (i.e. shorten it) in timeline.
Observe dot is gone from keyframe view (when shortening clip) or remains in place despite keyframe moving (when lengthening it)

Expected behaviour
Filters should follow keyframes when shortening or lenghtening clips.

Sorry if this description is unclear, I've been using Shotcut for all of one day now.  Searching the bugtracker didn't show up anything.

**Comment by ddennedy** [comment]
I do not completely understand.

> remains in place despite keyframe moving

How it can it be both remaining in place and moving?

> Filters should follow keyframes

What does that mean? Please restate that as "Keyframes should..."

Here is what I observe:
1. If you lengthen the clip, the keyframe remains at its current position.
2. If you shorten the clip, the keyframe remains at its current position unless...
3. If the clip becomes shorter than a keyframe's position that keyframe is removed.
4. If the clip duration is adjusted again to make it longer the removed keyframe does not reappear.

I think you are reporting about step 4 behavior.

**Comment by alexrudd2** [comment]
I think I misunderstood the GUI.  I thought the red line at the right of a clip was the keyframe, and the white dot was the end of a filter.  It appears the white dot is actually the keyframe. 

Step 4 is behavior is unexpected, yes.  In my opinion, keyframes should "follow" the end of the clip being moved.  Any filter parameters on this keyframe should keep their value as the keyframe moves.  (Alternately, a linear interpolation of their value).

It appears the effect is different with various filters.  I originally noticed on "Rotate and scale"  
1. Apply a clip filter of 100% scale at the beginning keyframe, and 20% at the end keyframe.  
2. The video shrinks as expected.  
3. Shorten the clip
4. The video disappears or shrinks only vertically??
There is no apparent way to "repair" the filter.  I tried toggling the keyframe button and changing the scale parameters at either keyframe.  The filter must be deleted and re-added.

(probably a separate bug, but it explains my confusion)
When I tested with the brightness filter, I noticed that the white dots are connected with a line, and their vertical position corresponds to the parameter value.  For the "Rotate and scale" filter, there is no line.  Also, the dots are all at the same position vertically regardless of value.

![keyframe_filter](https://user-images.githubusercontent.com/52292902/60297840-bce0d380-98ee-11e9-9c78-6a519caae4af.png)
![keyframe_filter_brightness](https://user-images.githubusercontent.com/52292902/60299245-e18a7a80-98f1-11e9-999d-bd47b40b0e69.png)

**Comment by ddennedy** [comment]
> I thought the red line at the right of a clip was the keyframe

Those bars are for trimming the filter - defining when it becomes active and deactive.

> keyframes should "follow" the end of the clip being moved

No, they do not by design. They are absolutely positioned because often times these keyframes are used for correction and are based on the state of the image or audio at that point in the clip. There is another feature you have not learned about called simple keyframes, and those are relative to the beginning or the end. I may some day add the option to make advanced keyframes relative to the beginning or the end of the clip.

 >  For the "Rotate and scale" filter, there is no line. Also, the dots are all at the same position vertically regardless of value.

This is by design. Not all parameters support the notion of curves because there is not a well defined range. I think they are not included with this filter's scale and offset parameters because they have and need a very large range thus making the curves a little difficult and odd for many use cases. 

In any case, I do not discuss design decisions with non-code-contributors, and it seems you more to learn. There is a keyframes video on our [Tutorials page](https://shotcut.org/tutorials/).

**Comment by alexrudd2** [comment]
You are correct, I am still learning the software.  The pan & zoom tutorial I watched was not the "official" keyframes one.  Apologies for not having the correct terminology.  I read your comments and the release notes, and still believe there is a bug here.  Specifically, filters misbehave when an ending keyframe is dropped by shortening a video.  (The behaviour when lengthening a video now makes sense to me.  It was hard to realize because I shortened first, thus "breaking" the filter)

As to not showing the curves for very large ranges, this humble software user suggests a logarithmic scale.
